pkg/storer/internal/reserve/reserve.go

```diff
@@ -97,7 +97,12 @@ func (r *Reserve) Put(ctx context.Context, chunk swarm.Chunk) error {
 	r.multx.Lock(string(chunk.Stamp().BatchID()))
 	defer r.multx.Unlock(string(chunk.Stamp().BatchID()))
 
-	has, err := r.Has(chunk.Address(), chunk.Stamp().BatchID())
+	batchHash, err := chunk.Stamp().Hash()
+	if err != nil {
+		return err
+	}
+
+	has, err := r.Has(chunk.Address(), chunk.Stamp().BatchID(), batchHash)
```

Comment:
Will this result in chunks that were re-stamped within the same batch as before appearing in the reserve twice?   An instance per stamp which includes the timestamp, since the updated timestamp changes the stamp's hash?

So instead of actually replacing the old stamp (or timestamp) with the new stamp (or timestamp), the new stamp is actually added to the reserves because the timestamp is updated?

Reply:
This concern should be verifiable in a test that inserts a stamped chunk into the reserve, updates the timestamp in the stamp (like a re-upload of a chunk would do), and then insert that updated timestamp chunk.   I suspect they'll both end up in the reserve rather than updating the timestamp.

Reply:
Here's the line in the original stamper where an existing stamp on a chunk has it's timestamp updated if it is re-stamped (re-uploaded) from the same batch.
https://github.com/ethersphere/bee/blob/58b62d6819fc40e2bb1e272c56a04b53ee06d0d6/pkg/postage/stamper.go#L51

Reply:
There wound't be duplicate chunks upon restamping. The test is covered by https://github.com/ethersphere/bee/blob/feat-add-batch-hash/pkg/storer/internal/reserve/reserve_test.go#L138

Note that `stampindex.LoadOrStore` does not check the stamp hash. So it'll return the old stampIndex (including the old hash) for the `batchID/stampIndex`.
We then remove the old chunk using this old stamp hash.
https://github.com/ethersphere/bee/blob/feat-add-batch-hash/pkg/storer/internal/reserve/reserve.go#L138

Reply:
> We then remove the old chunk using this old stamp hash. https://github.com/ethersphere/bee/blob/feat-add-batch-hash/pkg/storer/internal/reserve/reserve.go#L138

I have some (unproven) suspicions that `removeChunk` may introduce storage corruptions in the sharky.   I did a naive hack when trying to work around the non-updating timestamps and ended up with a totally ravaged localstore.  This was duplicated on several nodes and the only thing they had in common is that they were attempting to remove/re-insert chunks whose timestamp had changed.   I didn't call "foul" at that time because I always blame my own code first and it was more expeditious to just remove my hack and live with non-updating timestamps.   But this may need some good testing in the field, with database consistency checks run on nodes that have logged "`replacing chunk stamp index`".

Maybe even consider elevating this log to `Warning` since it should be rare enough and we'd really like to know it's happening even without `--verbosity 5`?

Reply:
Elevated to warning

Reply:
Hi @ldeffenb , bringing this back :)

Can you elaborate on your suspicion?
I'm seeing an issue where after updating a chunk timestamp, the old chunk is not removed from sharky (same address).

```
│ bee chunkstore: deleting chunk from sharky address 2a86d275607ef7cf284b15145c6539b1936128e7aa991b00100bec6fbcca86a8                                                                                                                                                                                                                                                                                           │
│ bee chunkstore: verifying delete on address 2a86d275607ef7cf284b15145c6539b1936128e7aa991b00100bec6fbcca86a8                                                                                                                                                                                                                                                                                                  │
│ bee chunkstore: failed to delete chunk from sharky address 2a86d275607ef7cf284b15145c6539b1936128e7aa991b00100bec6fbcca86a8                                                                                                                                                                                                                                                                                   │
│ bee "time"="2024-08-01 18:01:24.720744" "level"="warning" "logger"="node/storer/reserve" "msg"="replacing chunk stamp index" "old_chunk"="2a86d275607ef7cf284b15145c6539b1936128e7aa991b00100bec6fbcca86a8" "new_chunk"="2a86d275607ef7cf284b15145c6539b1936128e7aa991b00100bec6fbcca86a8"
│ bee "time"="2024-08-01 18:01:24.720864" "level"="info" "logger"="node/storer/reserve" "msg"="putting new chunk" "address"="2a86d275607ef7cf284b15145c6539b1936128e7aa991b00100bec6fbcca86a8" 
│ bee chunkstore: put chunk in sharky address 2a86d275607ef7cf284b15145c6539b1936128e7aa991b00100bec6fbcca86a8                                                                                                                                                                                                                                                                                                  │
│ bee chunkstore: chunk already exists in sharky. skipping address 2a86d275607ef7cf284b15145c6539b1936128e7aa991b00100bec6fbcca86a8 count 2  
```

https://github.com/ethersphere/bee/blob/bill/playground/pkg/storer/internal/chunkstore/chunkstore.go#L131-L137

Reply:
@acha-bill  My suspicion was caused by the observation that every node that I had that ended up with sharky corruption (detected by `db validate`) showed logs that indicated that a timestamp was updated causing `removeChunk` to be invoked during a Put operation.   I don't have direct evidence of what causes the corruption, but I suspect that somehow either an index is not removed properly or possibly not inserted properly.  But somehow, the sharky ended up with hashes that didn't match the associated index.

Reply:
https://github.com/ethersphere/bee/blob/bill/playground/pkg/storer/internal/chunkstore/chunkstore.go#L131-L137

Looking at this code block, remember that chunks don't necessarily leave the sharky on removal from just one index.   If a chunk is pinned in some reference, it will remain in because of a residual refCnt in the index.   But looking further back in the code, I see that was taken into account.

There may also be something in the actual sharky implementation that causes a retrieval to still work until something is committed thereby finalizing the delete.   It's been a long time since I looked at that code in detail.